backend/src/main/java/com/google/step/data/Lead.java

```diff
@@ -0,0 +1,235 @@
+    this.leadId = (String) entity.getProperty("leadId");
+    this.campaignId = (Long) entity.getProperty("campaignId");
+    this.gclId = (String) entity.getProperty("gclId");
+    this.apiVersion = (String) entity.getProperty("apiVersion");
+    this.formId = (Long) entity.getProperty("formId");
+    this.googleKey = (String) entity.getProperty("googleKey");
+    this.isTest = (Boolean) entity.getProperty("isTest");
+    this.adgroupId = (Long) entity.getProperty("adgroupId");
+    this.creativeId = (Long) entity.getProperty("creativeId");
+
+    entity.removeProperty("date");
+    entity.removeProperty("leadId");
+    entity.removeProperty("campaignId");
+    entity.removeProperty("gclId");
+    entity.removeProperty("apiVersion");
+    entity.removeProperty("formId");
+    entity.removeProperty("googleKey");
+    entity.removeProperty("isTest");
+    entity.removeProperty("adgroupId");
+    entity.removeProperty("creativeId");
+    this.columnData = new HashMap<>();
+    for (String key : entity.getProperties().keySet()) {
+      columnData.put(key, (String) entity.getProperty(key));
+    }
+  }
+
+  /**
+   * @return this object represented as JSON
+   */
+  public String asJson() {
+    return gson.toJson(this);
+  }
+
+  /**
+   * @return this lead object represented as an Entity
+   */
+  public Entity asEntity() {
+    Entity leadEntity = new Entity("Lead");
+    leadEntity.setProperty("date", date);
+    leadEntity.setProperty("leadId", leadId);
+    leadEntity.setProperty("campaignId", campaignId);
+    leadEntity.setProperty("gclId", gclId);
+    leadEntity.setProperty("apiVersion", apiVersion);
+    leadEntity.setProperty("formId", formId);
+    leadEntity.setProperty("googleKey", googleKey);
+    leadEntity.setProperty("isTest", isTest);
+    leadEntity.setProperty("adgroupId", adgroupId);
+    leadEntity.setProperty("creativeId", creativeId);
+    for (String key : columnData.keySet()) {
+      leadEntity.setProperty(key, columnData.get(key));
+    }
+    return leadEntity;
+  }
+
+  /**
+   * Creates a lead based off of JSON
+   * @param reader a reader object containing a JSON describing a lead object
+   * @return a lead object created by the JSON
+   */
+  public static Lead fromReader(Reader reader) {
+    Lead thisLead = gson.fromJson(reader, Lead.class);
+    thisLead.generateDataMap();
+    return thisLead;
+  }
+
+  /**
+   * Creates and populates the columnData Map from userColumnData
+   */
+  private void generateDataMap() {
+    columnData = new HashMap<>(userColumnData.size());
+    for (ColumnData cData : userColumnData) {
+      columnData.put(cData.getColumnId(), cData.getStringValue());
+    }
+  }
+
+  //Getters and Setters
+  /**
+   * @return the Date this lead was received
+   */
+  public Date getDate() {
+    return date;
+  }
+
+  /**
+   * @return current leadId
+   */
+  public String getLeadId() {
+    return leadId;
+  }
+
+  /**
+   * @return current campaignId
+   */
+  public Long getCampaignId() {
```

Comment:
can these use primitive long?

Reply:
We used the Long wrapper class in case the specified data value wasn't given to us by the lead JSON to allow for a null value instead of just a 0.

Reply:
Should we allow campaign id to be null? Is it possible for a lead to not belong to any campaign? If not, then we should return a 4xx status code (not retriable error) as shown here: https://screenshot.googleplex.com/jsS98n7cgwi

Reply:
Updated to primitive longs.